Uninstall modules

jvalleroy opened the pull request:
This adds an "Uninstall" button to the bottom of the page, for apps using the "service" view and template. The uninstall button won't be shown for essential services. When the user clicks "Uninstall", they are shown a confirmation page that lists all the packages that will be removed.

![plinth_uninstall1](https://cloud.githubusercontent.com/assets/178876/20032057/ae362134-a358-11e6-8fd8-11f2176f1d6b.png)

![plinth_uninstall2](https://cloud.githubusercontent.com/assets/178876/20032058/af38ec92-a358-11e6-9e5b-9ad1bb7164d0.png)

jvalleroy committed:
Add uninstall button for non-essential services

SunilMohanAdapa commented on plinth/service.py:
```diff
@@ -45,7 +45,8 @@ class Service(object):
     def __init__(self, service_id, name, ports=None, is_external=False,
-                 is_enabled=None, enable=None, disable=None, is_running=None):
+                 is_enabled=None, enable=None, disable=None, is_running=None,
+                 allow_uninstall=True):
```
allow_uninstall should not be property of a service but property of an application.  This is similar to how list of packages is itself a property of application instead of a property of a services.  The difference is that:
- Application has several services and all of them are installed and removed in an single step.
- Package may not be associated with any service.

SunilMohanAdapa commented on plinth/templates/service.html:
```diff
@@ -68,4 +68,18 @@
+      <br>
+      <p>
+        <a class="btn btn-default btn-md"
+           href="{% url 'uninstall' diagnostics_module_name %}">
```
We could later create a task to rename diagnostic_module_name to app_name and automatically infer it from Django application name.

SunilMohanAdapa commented on plinth/templates/uninstall.html:
```diff
@@ -0,0 +1,52 @@
+{% extends "base.html" %}
+{% comment %}
```
I don't think removing the left menu extra to give the uninstall process a completely different look is the best way to go. Since progress should be shown for uninstall, the user should be able to navigate to other parts of the Plinth during uninstall and be able to come back to see the progress. Also during uninstallation, how does the application page look like?

We should make the uninstall process look very similar to installation process to avoid these problems. After the uninstall conformation page is shown and the uninstall is happening, we can redirect the user to the application page where the progress is shown.

SunilMohanAdapa commented on plinth/views.py:
```diff
@@ -51,6 +54,31 @@ def index(request):
                              'details_label': details_label})
 
 
+def uninstall(request, modulename):
```
I believe this should be handled by the SetupView and SetupMiddleware.  The transaction handling part to should go to SetupHelper.

SunilMohanAdapa reviewed (changes requested):
Please see my comments and let me know what you think.

jvalleroy committed:
Drop allow_uninstall for services

Instead, modules that use ServiceView can specify that they are
essential modules.

jvalleroy committed:
Move uninstall url under /apps

Keep apps menu on the left side of uninstall page.

jvalleroy committed:
Use setup helper and middleware for uninstall process

Show uninstall view with progress for app, if uninstall process is ongoing.

Move uninstall routes into each app that supports uninstall.

jvalleroy commented:
@SunilMohanAdapa I made the following changes:
- Menu is shown on the left.
- Shows progress of uninstall process.
- If you go to the app page during the uninstall, it will show the progress view.

mray commented:
How big would be the implication to replace the current "disable" functionality by this?
The idea is to not have two different types of turning an app "off" (disable & uninstall).
I'd like to see a "proper" turning off as default, and "disable" as the optional extra.

JosephKiranBabu commented:
@jvalleroy What's the current thinking on this feature? Do we still think it's useful?
Though we have this story for clean (#973), I think uninstallation is useful as apt clean or autoremove doesn't do uninstallation.

mray commented:
I think we have to break the issue down from a users view to have only two states: ON/OFF.

The challenge is to decide on
**1. having sensible defaults**
**2. deciding what else might still be necessary, and where to add that.**

---

**1.**

One thing is for sure: we don't want to have the word "package" be of any relevance, that particularly includes explicit questions on their removal. We take care of dependency hell!
Any suggestions what should *actually* happen under the hood when asking: "Do you really want to turn this OFF?"

My take on defaults would be whatever `sudo apt-get remove` does.

**2.**
I can see how there is a dedicated configuration module that offers maintenance tasks like `apt-get clean` and `purge` for specific modules and whatever you may have to do when things get stuck (upgrade with -f flag??)...
If problems like "disk out of space" appear (we don't seem to check this yet by the way) we might link there for example.

JoKeyser commented:
My two cents:

1. I agree with @mray that "packages" could be avoided as a concept, but I **do** think it's fair to ask users to understand the difference between an (un)installed and en/disabled application. These are universal concepts in any computing device.
2. I think running `apt-get purge` might be a good idea, because removing old config file (formats) might be the only option to recover an App after failed upgrades etc...?

jvalleroy commented:
Repository moved to https://salsa.debian.org/freedombox-team/freedombox.
